### ISSUE TITLE
prevent generated proxy class to throw an exception in __toString while loading data

### ISSUE THREAD

**Issue by SimonHeimberg** [doctrine/common#824]
Handle the `__toString()` method special in generated proxy classes. It may not throw an error when loading the entity fails. Return some string instead.

### Reason

Throwing an exception in a __toString method triggers a fatal php error, aborting php (phpunit or whatever). When calling __toString() on a proxied entity which does not exist, this triggers an error. It is difficult to locate because a fatal error is triggered.

### how to fix

Could be fixed here
https://github.com/doctrine/common/blob/f8f359524f210710f11fe352ae23e0e655223c94/lib/Doctrine/Common/Proxy/ProxyGenerator.php#L761
or maybe in the initializer class.

The generated method should look similar to this:
```php
    public function __toString()
    {
        try {
            $this->__initializer__ && $this->__initializer__->__invoke($this, '__toString', []);
        } catch (\Exception $e) { // php5
            return 'ERROR: '.$e;
        } catch (\Throwable $e) { // php7
            return 'ERROR: '.$e;
        }

        return parent::__toString();
    }
```

Maybe the returned string should be configurable somehow. Or a special method `toStringErrorGetString($e)` could be called if it exists.

### versions

doctrine/common version: 2.7.3
php version: 7.0.22

**Comment by SimonHeimberg** [comment]
probably related to #303

**Comment by Ocramius** [comment]
Silencing it seems wrong to me: a fatal, while sub-optimal, is more
appropriate.

Marco Pivetta

http://twitter.com/Ocramius

http://ocramius.github.com/

**Comment by SimonHeimberg** [comment]
I agree that silencing is bad. But often the error would trigger somewhere near, and creates a usable backtrace. In THIS case I prefer silencing to not being able to locate the error source.

Suggestions for alternatives:
* We can write some debugging instructions in the doc.
* Maybe we can describe a workaround (is there one?) in the doc.
* We can vote for allowing __toString() to throw an error: https://bugs.php.net/bug.php?id=53648
* Can we somehow delay throwing the exception? (Before exit? Next use of object in case this ever happens... )
* ...

**Comment by Ocramius** [comment]
I think the instructions in the docs are a sufficientt addition.

Allowing `__toString()` to throw is something that isn't yet feasible in a safe way in the PHP engine.

**Comment by SimonHeimberg** [comment]
What could we write in the doc? I do not have a useful general trick how to debug to find out where this error occurred.

**Comment by Tobion** [comment]
I think this can be closed considering php 7.4 allows throwing exceptions in __toString.

**Comment by malarzm** [comment]
Agreed 👍
